### PR TITLE
Paying interest in chai to avoid the oracle misbehavior 

### PR DESCRIPTION
**Pull request by k1rill-fedoseev**
The following events were added:
* `TokensSwapped` event is emitted for each swap Dai => Chai, Chai => Dai
* `PaidInterest` event is emitted when interest is paid to interest receiver

For other changes, interest is now paid in Chai tokens. 
Added a receiver contract template with an example of Chai => Dai conversion logic on the receiver side.

### CODE REVIEW

File: contracts/upgradeable_contracts/InterestReceiver.sol [hunk 1]
```diff
@@ -0,0 +1,48 @@
+pragma solidity 0.4.24;
+
+import "../interfaces/IChai.sol";
+import "../interfaces/ERC677Receiver.sol";
+import "./Initializable.sol";
+import "./Ownable.sol";
+
```
Comment (akolotov) [inline contracts/upgradeable_contracts/InterestReceiver.sol]: Consider to add a script to deploy this contract and initialize the owner similar to `deploy/src/utils/deployERC20Token.js`

File: contracts/upgradeable_contracts/InterestReceiver.sol [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    function initialize(address _owner) external {
+        require(!isInitialized());
+        setOwner(_owner);
+        setInitialize();
+    }
+
+    /**
+    * @return Chai token contract address
+    */
+    function chaiToken() public view returns (IChai) {
```
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: The Dai contract address is stored in Chai contract, so I used it from there. Do you think it is better to hardcode it directly?

[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Yes, it also allows safe some gas for `withdraw` operation.

File: contracts/upgradeable_contracts/ChaiConnector.sol [hunk 3]
```diff
@@ -11,8 +11,14 @@ import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 * @dev This logic allows to use Chai token (https://github.com/dapphub/chai)
 */
 contract ChaiConnector is Ownable, ERC20Bridge {
+    // emitted when two tokens is swapped (e. g. Sai to Dai, Chai to Dai)
+    event TokensSwapped(address indexed from, address indexed to, uint256 value);
```
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/ChaiConnector.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/48fa3238dae38761582c3a7665f8b09f8b895158

File: contracts/upgradeable_contracts/ChaiConnector.sol [hunk 4]
```diff
@@ -156,23 +162,24 @@ contract ChaiConnector is Ownable, ERC20Bridge {
     * @dev Internal function for paying all available interest, in Dai tokens
     */
     function _payInterest() internal {
-        require(address(interestReceiver()) != address(0));
+        address receiver = address(interestReceiver());
```
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/ChaiConnector.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/69ebfba8df6339b091d39d0711b818d5dacc9800

File: contracts/upgradeable_contracts/ChaiConnector.sol [hunk 5]
```diff
@@ -156,23 +162,24 @@ contract ChaiConnector is Ownable, ERC20Bridge {
...
-        uint256 balanceBefore = daiBalance();
-        chaiToken().exit(address(this), chaiBalance().sub(investedAmountInChai()));
-        uint256 interestInDai = daiBalance().sub(balanceBefore);
 
         // solhint-disable-next-line not-rely-on-time
         uintStorage[LAST_TIME_INTEREST_PAID] = now;
 
-        erc20token().transfer(interestReceiver(), interestInDai);
+        uint256 interest = chaiBalance().sub(investedAmountInChai());
+        chaiToken().transfer(receiver, interest);
```
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/ChaiConnector.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/69ebfba8df6339b091d39d0711b818d5dacc9800

File: contracts/upgradeable_contracts/InterestReceiver.sol [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        require(_value == chaiToken().balanceOf(address(this)));
+
+        chaiToken().exit(address(this), _value);
+    }
+
+    /**
+    * @dev Withdraws DAI tokens from the receiver contract
+    * @param _to address of tokens receiver
+    */
+    function withdraw(address _to) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/e1a7c607e4251a18c9d96a6d18e5cb0ce8d4fa98

File: contracts/upgradeable_contracts/InterestReceiver.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/6449be868cfcd86b07dfe2d91b0d2e95aa2d4a10

File: contracts/upgradeable_contracts/InterestReceiver.sol [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import "../interfaces/IChai.sol";
+import "../interfaces/ERC677Receiver.sol";
+import "./Initializable.sol";
+import "./Ownable.sol";
+
+/**
+* @title InterestReceiver
+* @dev Example contract for receiving Chai interest and immediatly converting it into Dai
+*/
+contract InterestReceiver is ERC677Receiver, Initializable, Ownable {
```
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/6449be868cfcd86b07dfe2d91b0d2e95aa2d4a10

File: contracts/upgradeable_contracts/InterestReceiver.sol (hunk 1 above)
Comment (akolotov) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/9c0c9a2ce003096882558a424f7fc2be84f1de01

File: contracts/upgradeable_contracts/InterestReceiver.sol [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+    * @dev ERC677 transfer callback function, received interest from Chai token is converted into Dai and sent to owner
+    * @param _value amount of transferred tokens
+    */
+    function onTokenTransfer(address, uint256 _value, bytes) external returns (bool) {
+        require(isInitialized());
+        require(_value == chaiToken().balanceOf(address(this)));
+
+        chaiToken().exit(address(this), _value);
```
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Addressed by https://github.com/poanetwork/tokenbridge-contracts/pull/380/commits/6449be868cfcd86b07dfe2d91b0d2e95aa2d4a10

File: contracts/upgradeable_contracts/InterestReceiver.sol [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        return ERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
+    }
+
+    /**
+    * @dev ERC677 transfer callback function, received interest from Chai token is converted into Dai and sent to owner
+    * @param _value amount of transferred tokens
+    */
+    function onTokenTransfer(address, uint256 _value, bytes) external returns (bool) {
+        require(isInitialized());
+        require(_value <= chaiToken().balanceOf(address(this)));
```
Comment (akolotov) [inline contracts/upgradeable_contracts/InterestReceiver.sol]: I see that we can safe some gas if add a new variable will keep the result of `chaiToken().balanceOf(address(this))` there.

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol [hunk 10]
```diff
@@ -7,8 +7,6 @@ import "../../interfaces/IScdMcdMigration.sol";
 import "../ChaiConnector.sol";
 
 contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideBridgeStorage, ChaiConnector {
```
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Actually, it make sense. But, seems like Solc compiler does not allow it. I get the following:
```
Error: CompileError: .../ForeignBridgeErcToNative.sol:10:1: TypeError: Linearization of inheritance graph impossible
```
I think there is no need in introducing a one more separate interface for that, so we can keep it as it is. `TokenSwapper` is imported in `ChaiConnector`, and thus it is accessible in `ForeignBridgeErcToNative`.

File: deploy/src/loadEnv.js [hunk 11]
```diff
@@ -242,7 +243,15 @@ if (BRIDGE_MODE === 'ERC_TO_NATIVE') {
     BLOCK_REWARD_ADDRESS: addressValidator({
       default: ZERO_ADDRESS
     }),
-    FOREIGN_MIN_AMOUNT_PER_TX: bigNumValidator()
+    FOREIGN_MIN_AMOUNT_PER_TX: bigNumValidator(),
+    DEPLOY_INTEREST_RECEIVER: envalid.bool({default: false}),
```
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply deploy/src/loadEnv.js]: Yes, it should work like this.

File: contracts/upgradeable_contracts/InterestReceiver.sol (hunk 9 above)
Comment (k1rill-fedoseev) [reply contracts/upgradeable_contracts/InterestReceiver.sol]: Yes, you are right. Resolved by 642a24c

File: deploy/src/utils/deployInterestReceiver.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+} = require('../loadContracts')
+
+const {
+  DEPLOYMENT_ACCOUNT_PRIVATE_KEY,
+  FOREIGN_INTEREST_RECEIVER_OWNER
+} = env
+
+const DEPLOYMENT_ACCOUNT_ADDRESS = privateKeyToAddress(DEPLOYMENT_ACCOUNT_PRIVATE_KEY)
+
+async function deployInterestReceiver() {
```
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply deploy/src/utils/deployInterestReceiver.js]: Added a paragraph in README for that in 21225c5

File: deploy/.env.example [hunk 13]
```diff
@@ -70,6 +70,10 @@ DPOS_STAKING_ADDRESS=
 #If decimals are different between foreign and home chain, number of decimals shift required to adjust the amount of tokens bridged. Default value is 0.
 FOREIGN_TO_HOME_DECIMAL_SHIFT=
 
+# for bridge erc_to_native mode
```
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply deploy/.env.example]: Added a brief parameter description, moved them to the end of the file. 21225c5

**Comment by patitonar** [comment]
I tested the upgrade procedure in Sokol-Kovan and it worked correctly, no issues were found. Also, I didn't found any issues with Chai integration. The issue found in the [previous test](https://github.com/poanetwork/tokenbridge-contracts/pull/379#issuecomment-592645589) is now fixed.

HomeBridge: [0x4748C57C307082Fb473f8E140ac4D3232166329e](https://blockscout.com/poa/sokol/address/0x4748C57C307082Fb473f8E140ac4D3232166329e)

ForeignBridge: [0x29bB8B94214197DF0486211E74f1AE41e4C41cc7](https://kovan.etherscan.io/address/0x29bB8B94214197DF0486211E74f1AE41e4C41cc7)

Multisig as foreign owner to test the upgrade scripts: [0x6829Df9a24D46f8C2fa22b9c59f9cf30FF49206a](https://kovan.etherscan.io/address/0x6829Df9a24D46f8C2fa22b9c59f9cf30FF49206a)

Chai: [0x623f148dabd40dd221807c5e98235bc40e54a106](https://kovan.etherscan.io/token/0x623f148dabd40dd221807c5e98235bc40e54a106) (Used the deployed Chai token mentioned [here](https://github.com/poanetwork/tokenbridge-contracts/pull/356#issuecomment-577648189))

Here is the test performed:

0. Start with the current implementation of contract and oracle
1. Sent some transactions on both sides
2. Deployed the new implementation contract in foreign.
3. Upgrade oracle.
4. Sent a transaction from Foreign to home and worked correctly. Sent a transaction from Home to Foreign and it failed at finishing processing it as expected because of how signatures are submitted.
5. Upgrade Foreign contract with new implementation using the script `upgradeBridgeOnForeign`. First one validator as `leader` role and the other two as `confirm`. Here I set `NEW_IMPLEMENTATION_ETH_BRIDGE=0xBD405611232A30a77619Fc731438920d147bE929`
6. Pending transfer from Home to Foreign completed correctly.
7. Sent transactions in both directions.
8. Initialized Chai and receiver using the script `initializeChai`. First one validator as `leader` role and the other two as `confirm`. Here I set `CHAI_INTEREST_RECEIVER=0xf4BEF13F9f4f2B203FAF0C3cBbaAbe1afE056955`(EOA).
9. Sent some transactions on both sides. Dai tokens were converted to Chai.
10. Waited some time to generate interest.
11. Called `payInterest()` method. [Here is the Tx](https://kovan.etherscan.io/tx/0x97df9ad952a3cf1aac45e2c1a3c7261ba578dfe63fa982fbd11a291ef51cc418).
12. Tested running and processing transactions from an oracle instance deployed with the deployment script.

**Pull request merged**